renku/ui/service/gateways/repository_cache.py

```diff
@@ -0,0 +1,233 @@
+
+import os
+import shutil
+import uuid
+from datetime import datetime
+from typing import Optional
+from urllib.parse import urlparse
+
+import portalocker
+from marshmallow import ValidationError
+
+from renku.command.clone import project_clone_command
+from renku.core import errors
+from renku.core.util.contexts import renku_project_context
+from renku.core.util.os import normalize_to_ascii
+from renku.domain_model.git import GitURL
+from renku.infrastructure.repository import Repository
+from renku.ui.service.cache import ServiceCache
+from renku.ui.service.cache.models.project import Project
+from renku.ui.service.cache.models.user import User
+from renku.ui.service.config import PROJECT_CLONE_DEPTH_DEFAULT
+from renku.ui.service.errors import IntermittentCacheError, IntermittentLockError
+from renku.ui.service.interfaces.repository_cache import IRepositoryCache
+from renku.ui.service.logger import service_log
+
+
+class LocalRepositoryCache(IRepositoryCache):
+    """Cache for project repos stored on local disk."""
+
+    def get(
+        self, cache: ServiceCache, git_url: str, branch: Optional[str], user: User, shallow: bool = True
+    ) -> Project:
+        """Get a project from cache (clone if necessary)."""
+        if git_url is None:
+            raise ValidationError("Invalid `git_url`, URL is empty", "git_url")
+
+        try:
+            project = Project.get(
+                (Project.user_id == user.user_id) & (Project.git_url == git_url) & (Project.branch == branch)
+            )
+        except ValueError:
+            # project not found in DB
+            return self._clone_project(cache, git_url, branch, user, shallow)
+
+        if not project.abs_path.exists():
+            # cache folder doesn't exist anymore
+            project.delete()
+            return self._clone_project(cache, git_url, branch, user, shallow)
+
+        if not shallow and project.clone_depth is not None and project.clone_depth <= 1:
+            self._unshallow_project(project, user)
+
+        self._maybe_update_cache(project, user)
+
+        if not project.initialized:
+            raise errors.UninitializedProject(project.git_url)
+
+        self._update_project_access_date(project)
+
+        return project
+
+    def evict(self, project: Project):
+        """Evict a project from cache."""
+        try:
+            with project.write_lock():
+                service_log.debug(f"purging project {project.project_id}:{project.name}")
+                project.purge()
+        except FileNotFoundError:
+            project.delete()
+        except Exception as e:
+            service_log.error(f"Couldn't purge project {project.project_id}:{project.name} from cache", exc_info=e)
+
+    def evict_expired(self):
+        """Evict expired projects from cache."""
+        for project in Project.all():
+            if project.ttl_expired():
+                self.evict(project)
+
+    def _update_project_access_date(self, project: Project):
```

Comment:
Just an optional suggestion... Is this something that should be a decorator used on other methods here so that it updates the accessed date when specific methods are called?

Reply:
Ok sounds good. I thought it may not work. Just thought I'd bring it up.